starboard/shared/libiamf/iamf_buffer_parser.cc

```diff
@@ -0,0 +1,588 @@
+    case kFourccOpus:
+      info->sample_rate = kOpusSampleRate;
+      break;
+    case kFourccIpcm: {
+      // sample_format_flags
+      RCHECK(reader->SkipBytes(1));
+
+      // sample_size
+      RCHECK(reader->SkipBytes(1));
+
+      uint32_t sample_rate_unsigned;
+      RCHECK(reader->Read4(&sample_rate_unsigned));
+      info->sample_rate = static_cast<int>(sample_rate_unsigned);
+      break;
+    }
+    default:
+      SB_NOTREACHED();
+      return false;
+  }
+
+  return true;
+}
+
+bool IamfBufferParser::ParseAudioElementOBU(BufferReader* reader,
+                                            IamfBufferInfo* info,
+                                            const bool prefer_binaural_audio,
+                                            const bool prefer_surround_audio) {
+  uint32_t audio_element_id;
+  RCHECK(reader->ReadLeb128(&audio_element_id));
+
+  uint8_t audio_element_type;
+  RCHECK(reader->Read1(&audio_element_type));
+  audio_element_type = audio_element_type >> 5;
+
+  // codec_config_id
+  RCHECK(reader->SkipLeb128());
+
+  uint32_t num_substreams;
+  RCHECK(reader->ReadLeb128(&num_substreams));
+
+  for (int i = 0; i < num_substreams; ++i) {
+    // audio_substream_id
+    RCHECK(reader->SkipLeb128());
+  }
+
+  uint32_t num_parameters;
+  RCHECK(reader->ReadLeb128(&num_parameters));
+
+  for (int i = 0; i < num_parameters; ++i) {
+    uint32_t param_definition_type;
+    RCHECK(reader->ReadLeb128(&param_definition_type));
+
+    if (param_definition_type == IAMF_PARAMETER_TYPE_DEMIXING) {
+      SkipParamDefinition(reader);
+      // DemixingParamDefintion
+      RCHECK(reader->SkipBytes(1));
+    } else if (param_definition_type == IAMF_PARAMETER_TYPE_RECON_GAIN) {
+      SkipParamDefinition(reader);
+    } else if (param_definition_type > 2) {
+      uint32_t param_definition_size;
+      RCHECK(reader->ReadLeb128(&param_definition_size));
+      RCHECK(reader->SkipBytes(param_definition_size));
+    }
+  }
+
+  if (static_cast<uint32_t>(audio_element_type) ==
+          AUDIO_ELEMENT_CHANNEL_BASED &&
+      (prefer_binaural_audio || prefer_surround_audio)) {
+    // Parse ScalableChannelLayoutConfig for binaural and surround
+    // loudspeaker layouts
+    uint8_t num_layers;
+    RCHECK(reader->Read1(&num_layers));
+    num_layers = num_layers >> 5;
+    // Read ChannelAudioLayerConfigs
+    for (int i = 0; i < static_cast<int>(num_layers); ++i) {
+      uint8_t loudspeaker_layout;
+      bool output_gain_is_present_flag;
+      RCHECK(reader->Read1(&loudspeaker_layout));
+      output_gain_is_present_flag = (loudspeaker_layout >> 3) & 0x01;
+      loudspeaker_layout = loudspeaker_layout >> 4;
+      if (loudspeaker_layout == IA_CHANNEL_LAYOUT_BINAURAL) {
+        binaural_audio_element_ids_.insert(audio_element_id);
+      } else if (loudspeaker_layout > IA_CHANNEL_LAYOUT_STEREO &&
+                 loudspeaker_layout < IA_CHANNEL_LAYOUT_COUNT) {
+        surround_audio_element_ids_.insert(audio_element_id);
```

Comment:
I suggest to use one set for both binaural and surround audio elements, and store both element id and layout info in the set, like std::unordered_set<std::pari<uint32_t, uint8_t>>.